src/lib/zork/zork.ts

```diff
@@ -0,0 +1,387 @@
+  'getters': handleCmdGetters,
+  'transform': handleCmdTransform,
+  'give': handleCmdGive,
+  'get': handleCmdGet
+};
+
+
+const handleSignaling = (ctx: Context, msg: string) => {
+  const data = JSON.parse(msg);
+  if (data.type === 'icecandidate' && data.candidate) {
+    handleIce(ctx, data);
+  } else if (data.type === 'offer' && data.sdp && ctx.mode === 'give') {
+    // Getter is offerer, so giver receives the offer.
+    ctx.rtc.remoteReceived = true;
+    ctx.log(`got offer: ${msg} -> remoteReceived = true`);
+    numGetters++;
+    ctx.log(`incremented numGetters to ${numGetters}`);
+    ctx.rtc.conn.setRemoteDescription(data, () => {
+      while (ctx.rtc.pendingCandidates.length) {
+        ctx.rtc.conn.addIceCandidate(ctx.rtc.pendingCandidates.shift());
+      }
+      ctx.rtc.conn.createAnswer((answer: any) => {
+        ctx.log(`created answer, setting local description`);
+        ctx.rtc.conn.setLocalDescription(answer, () => {
+          const json = JSON.stringify(answer);
+          ctx.log(`forwarding answer: ${json}`);
+          ctx.reply(json);
+        }, ctx.log.error);
+      }, ctx.log.error);
+    }, ctx.log.error);
+  } else if (data.type === 'answer' && data.sdp && ctx.mode === 'get') {
+    // Giver is answerer, so getter receives the answer.
+    ctx.rtc.remoteReceived = true;
+    ctx.log(`got answer: ${msg} -> remoteReceived = true`);
+    ctx.rtc.conn.setRemoteDescription(data);
+  } else {
+    ctx.log(`handleSignaling: ignoring msg: ${msg}`);
+  }
+};
+
+const handleIce = (ctx: Context, data: any) => {
+  if (ctx.rtc.remoteReceived) {
+    ctx.log(`handleIce: remoteReceived: adding ice candidate...`);
+    ctx.rtc.conn.addIceCandidate(data.candidate);
+  } else {
+    ctx.log(`handleIce: !remoteReceived: adding pending ice candidate`);
+    ctx.rtc.pendingCandidates.push(data.candidate);
+  }
+};
+
+
+const parseCmd = (cmdline: string) : ParsedCmd => {
+  const tokens = cmdline.split(/\W+/);
+  const verb = tokens[0].toLowerCase();
+  const parsed = {verb: verb, tokens: tokens, source: cmdline};
+  return parsed;
+};
+
+
+const handleMsg = (ctx: Context, msg: string) => {
+  if (ctx.mode) { // Already in 'give' or 'get' mode. Treat msg as signaling.
+    handleSignaling(ctx, msg);
+  } else { // Not yet in 'give' or 'get' mode. Treat msg as command.
+    const cmd = parseCmd(msg);
+    const cmdHandler = cmdHandlerByVerb[cmd.verb] || handleCmdInvalid;
+    cmdHandler(ctx, cmd);
+  }
+};
+
+
+const onSocksConnection = (ctx: Context, sessionId: any) => {
+  ctx.log(`[socksServer] new connection: ${sessionId}`);
+  const channel = ctx.rtc.conn.createDataChannel(sessionId);
+  return {
+    // SOCKS client -> datachannel
+    handleDataFromSocksClient: (bytes: ArrayBuffer) => channel.send(bytes),
+    // SOCKS client <- datachannel
+    onDataForSocksClient: (callback: (buffer: ArrayBuffer) => void) => {
+      channel.onmessage = (event: any) => callback(event.data);
+      return this;
+    },
```

Comment:
Here in onSocksConnection we create a new datachannel, but we can't set its onmessage handler right away (as we can from the "give" side), because the SocksServer currently doesn't expose a method we could call from the onmessage handler with the event.data (whereas on the give side SockSession does expose handleMessageFromSocksClient). So instead we currently have to supply an onDataForSocksClient callback, which itself takes a callback, and only in our onDataForSocksClient callback do we set our channel.onmessage handler to a function that calls the supplied callback.

Would it be possible to refactor this to take away one level of callback inversion, the way it works on the give side? I may have a go at this if so.